// @remotion/core/dist/sequencing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sequence = exports.SequenceContext = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const use_frame_1 = require("../use-frame");
exports.SequenceContext = react_1.createContext(null);
const Sequence = ({ from, durationInFrames: duration, children }) => {
    const currentFrame = use_frame_1.useAbsoluteCurrentFrame();
    const contextValue = react_1.useMemo(() => {
        return {
            from,
            durationInFrames: duration,
        };
    }, [duration, from]);
    return (jsx_runtime_1.jsx(exports.SequenceContext.Provider, Object.assign({ value: contextValue }, { children: jsx_runtime_1.jsx("div", Object.assign({ style: {
                position: 'absolute',
                display: 'flex',
                width: '100%',
                height: '100%',
                top: 0,
                bottom: 0,
                left: 0,
                right: 0,
            } }, { children: currentFrame < from
                ? null
                : currentFrame > from + duration
                    ? null
                    : children }), void 0) }), void 0));
};
exports.Sequence = Sequence;
